wgpu/src/api/command_encoder.rs

```diff
@@ -346,4 +346,86 @@ impl CommandEncoder {
+    /// * CommandBuffer C: Use resources X and Y in a bind group
+    ///
+    /// To prevent this, after profiling their app, an advanced user might choose to instead do `queue.submit(&[a, b, c])`:
+    /// * CommandBuffer A:
+    ///     * Use [`CommandEncoder::transition_resources`] to transition resources X and Y from TextureUses::RESOURCE (from last frame) to TextureUses::COLOR_TARGET
+    ///     * Use resource X as a render pass attachment
+    /// * CommandBuffer B: Use resource Y as a render pass attachment
+    /// * CommandBuffer C:
+    ///     * Use [`CommandEncoder::transition_resources`] to transition resources X and Y from TextureUses::COLOR_TARGET to TextureUses::RESOURCE
+    ///     * Use resources X and Y in a bind group
+    ///
+    /// At submission time, wgpu will record and insert some new command buffers, resulting in a submission that looks like `queue.submit(&[0, a, b, 1, c])`:
+    /// * CommandBuffer 0: Barrier to transition resources X and Y from TextureUses::RESOURCE (from last frame) to TextureUses::COLOR_TARGET
+    /// * CommandBuffer A: Use resource X as a render pass attachment
+    /// * CommandBuffer B: Use resource Y as a render pass attachment
+    /// * CommandBuffer 1: Barrier to transition resources X and Y from TextureUses::COLOR_TARGET to TextureUses::RESOURCE
+    /// * CommandBuffer C: Use resources X and Y in a bind group
+    ///
+    /// Which eliminates the extra command buffer and barrier between command buffers A and B.
+    ///
+    /// # Native Interoperability
+    ///
+    /// A user wanting to interoperate with the underlying native graphics APIs (Vulkan, DirectX12, Metal, etc) can use this API to generate barriers between wgpu commands and
+    /// the native API commands, for synchronization and resource state transition purposes.
+    #[cfg(wgpu_core)]
+    pub fn transition_resources(
+        &mut self,
+        buffer_transitions: &[BufferTransition<'_>],
+        texture_transitions: &[TextureTransition<'_>],
+    ) {
+        if let Some(encoder) = self.inner.as_core_mut_opt() {
+            encoder.transition_resources(buffer_transitions, texture_transitions);
+        }
+    }
+}
+
+/// A buffer transition for use with [`CommandEncoder::transition_resources`].
+#[cfg(wgpu_core)]
+#[derive(Debug)]
+pub struct BufferTransition<'a> {
+    /// The buffer to transition.
+    pub buffer: &'a Buffer,
+    /// The new state to transition to.
+    pub state: hal::BufferUses,
+}
+
+/// A texture transition for use with [`CommandEncoder::transition_resources`].
+#[cfg(wgpu_core)]
+#[derive(Debug)]
+pub struct TextureTransition<'a> {
+    /// The texture to transition.
+    pub texture: &'a Texture,
+    /// An optional selector to transition only part of the texture.
+    ///
+    /// If None, the entire texture will be transitioned.
+    pub selector: Option<wgc::TextureSelector>,
+    /// The new state to transition to.
+    pub state: hal::TextureUses,
```

Comment:
I feel decently strongly that we probably should move these into wgpu types so this api is cross platform not just wgpu-core. The webgpu impl would be a no-op. This is harder to use because you need to start digging into wgc/hal instead of just exposing this to the user normally as part of the interface.

Reply:
I had that initially, but it's a lot more of a mess, and dosen't really seem worth it to me given that the function kinda depends on wgpu-core implementation details anyways.

Reply:
Discussing this at next meeting.

Reply:
Trying again this meeting :)

Reply:
Alright, the resolution here from the meeting is that we should not have wgc/hal exposed in this api, we should move both of these types to wgpu-types, rewire core and hal. `BufferUses`/`TextureUses` due to the similarity to `BufferUsages`/`TextureUsages`, should be right next to their counterparts and have comments explaining why you'd want to use one over the other.

For maintainership purposes, this move should be done as a single commit, so it can be easy to review.